### PR TITLE
strategy: improve harmonic by adding HMM filter to denoise shark signal

### PR DESCRIPTION
**Pull request by austin362667**
strategy: improve harmonic by adding [HMM](https://tr8dr.github.io/HMMFiltering/) filter to denoise shark signal

### CODE REVIEW

**Review by andycheng123 — APPROVED** [review]

File: pkg/strategy/harmonic/strategy.go [hunk 1]
```diff
@@ -332,66 +336,131 @@ func (s *Strategy) Run(ctx context.Context, orderExecutor bbgo.OrderExecutor, se
...
+func transitionProbability(x_t0, x_t1 int) float64 {
+	// stick to the same sate
+	if x_t0 == x_t1 {
+		return 0.99
+	}
+	// transit to next new state
+	return 1 - 0.99
+}
+
+func alpha(y_t []float64, x_t []float64, l int) float64 {
```
Comment (c9s) [inline pkg/strategy/harmonic/strategy.go:429]: can you add comment for this alpha function?

File: pkg/strategy/harmonic/strategy.go [hunk 2]
```diff
@@ -332,66 +336,131 @@ func (s *Strategy) Run(ctx context.Context, orderExecutor bbgo.OrderExecutor, se
...
+		return 1.
+	} else if x_t < 0. && y_t < 0. {
+		// observed negative value from indicator when in short state
+		return 1.
+	} else {
+		return 0.
+	}
+}
+
+func transitionProbability(x_t0, x_t1 int) float64 {
```
Comment (c9s) [inline pkg/strategy/harmonic/strategy.go:420]: call it `transitProbability`?

File: pkg/strategy/harmonic/strategy.go [hunk 3]
```diff
@@ -332,66 +336,131 @@ func (s *Strategy) Run(ctx context.Context, orderExecutor bbgo.OrderExecutor, se
...
+
+		_, _ = fmt.Fprintln(os.Stderr, s.TradeStats.String())
+		_ = s.orderExecutor.GracefulCancel(ctx)
+	})
+
 	return nil
 }
+
+// TODO: dirichlet distribution is a too naive solution
+func observationDistribution(y_t, x_t float64) float64 {
```
Comment (c9s) [inline pkg/strategy/harmonic/strategy.go:405]: observeDistribution ?

File: pkg/strategy/harmonic/strategy.go [hunk 4]
```diff
@@ -332,66 +336,131 @@ func (s *Strategy) Run(ctx context.Context, orderExecutor bbgo.OrderExecutor, se
...
-			}
-			_, err := s.orderExecutor.SubmitOrders(ctx, types.SubmitOrder{
+		} else if states.Mean(5) == -1 && direction != -1 {
+			_, _ = s.orderExecutor.SubmitOrders(ctx, types.SubmitOrder{
 				Symbol:   s.Symbol,
 				Side:     types.SideTypeSell,
 				Quantity: s.Quantity,
 				Type:     types.OrderTypeMarket,
-				Tag:      "shark short: sell in",
+				Tag:      "shark short",
```
Comment (c9s) [inline pkg/strategy/harmonic/strategy.go:378]: Update tag to `sharkShort`

File: pkg/strategy/harmonic/strategy.go [hunk 5]
```diff
@@ -332,66 +336,131 @@ func (s *Strategy) Run(ctx context.Context, orderExecutor bbgo.OrderExecutor, se
 	if klines, ok := kLineStore.KLinesOfInterval(s.shark.Interval); ok {
 		s.shark.LoadK((*klines)[0:])
 	}
+
+	states := types.NewQueue(s.Window)
+
+	states.Update(0)
 	s.session.MarketDataStream.OnKLineClosed(types.KLineWith(s.Symbol, s.Interval, func(kline types.KLine) {
 
 		log.Infof("Shark Score: %f, Current Price: %f", s.shark.Last(), kline.Close.Float64())
```
Comment (c9s) [inline pkg/strategy/harmonic/strategy.go:345]: logs should use lower case

File: pkg/strategy/harmonic/strategy.go [hunk 6]
```diff
@@ -3,14 +3,17 @@ package harmonic
...
+
 	"github.com/c9s/bbgo/pkg/bbgo"
 	"github.com/c9s/bbgo/pkg/data/tsv"
 	"github.com/c9s/bbgo/pkg/datatype/floats"
 	"github.com/c9s/bbgo/pkg/fixedpoint"
 	"github.com/c9s/bbgo/pkg/indicator"
 	"github.com/c9s/bbgo/pkg/types"
-
 	"github.com/sirupsen/logrus"
+	floats2 "gonum.org/v1/gonum/floats"
```
Comment (c9s) [inline pkg/strategy/harmonic/strategy.go:16]: we have 2 floats?

**Comment by austin362667** [comment]
@c9s  multiple fixes done

**Review by c9s — APPROVED** [review]

**Pull request merged**